cmd/mirror-main.go

```diff
@@ -282,6 +289,38 @@ func (mj *mirrorJob) doMirror(ctx context.Context, cancelMirror context.CancelFu
 		sURLs.TargetContent.Metadata["X-Amz-Storage-Class"] = mj.storageClass
 	}
 
+	if mj.isPreserve {
+		// To determine the source if file system or object storage
+		// hostCfgSource == nil implies that the source is file system
+		_, _, hostCfgSource, err := expandAlias(sURLs.SourceAlias)
+		fatalIf(err, "Unable to initialize `"+sURLs.TargetAlias+"`.")
+
+		_, _, hostCfgTarget, err := expandAlias(sURLs.TargetAlias)
+		fatalIf(err, "Unable to initialize `"+sURLs.TargetAlias+"`.")
+
+		// when source is objectstorage
+		if hostCfgSource != nil {
+			statURL := sURLs.SourceAlias + getKey(sURLs.SourceContent)
+			_, sourceContent, _ := url2Stat(statURL, true, nil)
+			for key, value := range sourceContent.Metadata {
+				// when target is file system set mc-attrs
+				if hostCfgTarget == nil {
+					// if source has file system attribute as metadata
+					if key == "X-Amz-Meta-Mc-Attrs" {
+						sURLs.TargetContent.Metadata["mc-attrs"] = value
+					} else if key == "X-Amz-Meta-S3cmd-Attrs" {
+						sURLs.TargetContent.Metadata["mc-attrs"] = value
+					}
+				}
+			}
+		} else {
```

Comment:
This is causing abstraction leak we shouldn't do these things, you should enhance the code such that a filesystem Stat returns appropriate attributes and as well as object storage should also differentiate. 

In this code there are two different paths taken showing the actual differentiation which is exactly we should avoid and have a single call.

You need to redo this PR @sinhaashish